Rewrite for playground features and browser support

Devon7925 opened the pull request:
closes #9
closes #24
closes #30

This PR adds the following:
- Browser support using slang-wasm's version of slangd
- The ability to run playground shaders in the editor by opening a webview. A second webview is opened for uniforms if the shader uses them.
- The ability to compile to the same set of targets the playground supports
- The ability to get reflection-json information
- User accessible documentation for playground features.

Still TODO:
- [x] Currently `playground.slang` is inserted into every folder in the web version so that slang can recognize the import. This is not done in the native version, and I'm not sure how to do it.
  - I added a quickfix to create playground.slang if the import is failing. This isn't a great solution since if the playground updates the file will be outdated.
- [x] The current spirv-tools wasm build doesn't work on native, so the SPIR-V compile doesn't work. This should be an easy fix.
- [x] Compilation and deploys for different native platforms are not yet integrated.
- [x] Native multi file is not currently tested (and is likely broken)
- [x] I suspect playground specific features should not always be enabled, not sure what logic should be used though.
  - Run button is now hidden for files without `imageMain` or `printMain`. The other buttons can be used for any file or can be hidden by the user.
- [x] Currently pressing run while a shader is already running creates a second shader window. It should either take over or shut down the old one. Possibly runs from different files should be kept?
- Currently the web language server supports [virtual workspaces](https://code.visualstudio.com/api/extension-guides/virtual-workspaces), but the native language server does not. Ideally there would be a fallback mechanism but I'm considering it out of scope for this pr.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/shader-slang/slang-vscode-extension?pullRequest=35) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/shader-slang/slang-vscode-extension?pullRequest=35) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/shader-slang/slang-vscode-extension?pullRequest=35) it.</sub>

csyonghe commented on client/package-lock.json:
```diff
@@ -0,0 +1,2273 @@
+{
```
should this file be ignored?

Devon7925 replied:
`package.json` alone specifies a range of versions to pull, if the file isn't included it's possible a newer version of a dependency can be pulled when CI is ran that has an issue that breaks something, leading to an issue that happens only in deploy but not in local development.

Devon7925 replied:
That being said, I understand it makes commit diffs large and the issue is an edge case. If you think I should remove them I will.

csyonghe commented on client/package.json:
```diff
@@ -0,0 +1,23 @@
+{
+    "name": "devon-slang-playground-extension",
```
rename this to slang-playground-extension?

csyonghe commented on server/package.json:
```diff
@@ -0,0 +1,21 @@
+{
+    "name": "devon-slang-playground-extension",
```
name and repo url should be updated.

csyonghe commented on server/package-lock.json:
```diff
@@ -0,0 +1,121 @@
+{
```
this shouldn't be checked in.

csyonghe commented:
I guess with this, we can now get intellisense on github.dev when editing slang shaders?

Devon7925 commented on client/src/browserClientMain.ts:
```diff
@@ -0,0 +1,60 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
```
Same reason the current `extension.ts` is copyright Microsoft 😆. I've removed it.

Devon7925 commented:
Yeah, still won't be as good as using slangd on native, but it'll be an improvement.

natevm commented:
I mentioned it with Devon over the discord, we have a memory leak somewhere with this extension that we need to find before this gets merged in. 

![image](https://github.com/user-attachments/assets/cde5297b-41b6-480b-8a7a-205ad72c3870)

Other than that, I really like the extension so far.

natevm commented:
No, not on an insiders build. I’m also using the version on marketplace. Are there changes here which aren’t reflected over in the marketplace version?

Devon7925 commented:
Yes, the marketplace version is from before I made the changes to merge it into the main extension, and it uses the browser entrypoint even on native. Try using a build from here https://github.com/Devon7925/slang-vscode-extension/actions/runs/16081082875.

natevm commented:
> Yes, the marketplace version is from before I made the changes to merge it into the main extension, and it uses the browser entrypoint even on native. Try using a build from here https://github.com/Devon7925/slang-vscode-extension/actions/runs/16081082875.

Fortunately I don't seem to run out of memory with this more recent artifact. It must have been some bug with a previous iteration.

natevm commented:
I used the extension all through yesterday and never hit the OOM issue. I did hit a couple other UI issues. Perhaps these would be better as followup PRs though…

1. When importing a slang module, eg some utils.slang, if that utils.slang has some code which does not compile, the extension throws an unknown error rather than telling me a compilation error occurred.
2. In a couple cases, only the compile and reflection buttons were visible, and the play button was missing. Usually flipping between tabs and compiling once to WGSL seems to fix the issue. I could just be “holding it wrong”
3. When a VSCode terminal is connected over SSH to a remote machine (which happens a lot at NV) the extension throws an error and doesn’t work.
4. Syntax highlighting seems a bit more hit or miss. In many cases, it gets stuck highlighting completely valid text as though there were some compilation error there. I suspect certain Unicode characters might be the issue, but I haven’t been able to reliably repro the issue. It also seems to happen on the slang playground website itself.
5. The fullscreen button on the bottom right corner of the canvas does not work. Perhaps we should hide it or try to fix it?

Devon7925 commented:
I've addressed 2, 5, and 6. Not exactly sure what you're referring to with 1, but hopefully the improved logging I added helps with that. I agree 3 and 4 can be handled in follow-up PRs.

csyonghe reviewed (approved):
Looks good to me, thanks for the awesome work!

csyonghe commented on package.json:
```diff
@@ -29,8 +29,56 @@
+                "command": "slang.playgroundRun",
+                "title": "Run Playground",
+                "category": "Slang",
+                "enablement": "editorLangId == slang",
```
Is it possible to associate a default hot key to this, e.g. F5?

csyonghe replied:
Would be nice if we can interpretable shader files and run `slangi` on the file if the file contains a `void main()` or `int main()`.

Devon7925 replied:
I'm not familiar with `slangi`. Are docs for it somewhere?

csyonghe commented on package.json:
```diff
@@ -29,8 +29,56 @@
+                "command": "slang.compile",
+                "title": "Slang Compile",
+                "category": "Slang",
+                "enablement": "editorLangId == slang",
```
default hotkey this to `Ctrl+B`?

Devon7925 replied:
There are currently default hotkeys for run and compile:
```json
{
    "command": "slang.compile",
    "key": "ctrl+shift+b",
    "when": "editorLangId == slang"
},
{
    "command": "slang.playgroundRun",
    "key": "ctrl+shift+r",
    "when": "editorLangId == slang"
}
```
Do you want them changed to `f5` and `Ctrl+B`?

csyonghe commented on uniform_webview/package.json:
```diff
@@ -0,0 +1,16 @@
+{
```
I see many different package.json files in this PR, what's their purpose?

Devon7925 replied:
They're for different js builds. Theres one each for:
* The main extension build
* The "server" build used for interfacing with slang
* The main webview that renders the playground run
* The uniform components webview
* The files that are shared across multiple builds
* The extension itself to give vscode information

Since they have different dependencies/environments they get different package.json files.

csyonghe commented on webview/src/RenderCanvas.vue:
```diff
@@ -0,0 +1,1031 @@
+<script setup lang="ts">
```
I wonder if it is possible to reuse these components from both playgroud website and this extension.

Devon7925 replied:
It should be, though it'll require some extra work and probably a separate repository. Do you want me to work on that next?
